### PR TITLE
feat(calling): join subconversation

### PR DESCRIPTION
**Pull request by typfel**
----
#### PR Submission Checklist for internal contributors

- The **PR Title**
  - [ ] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
  - [ ] contains a reference JIRA issue number like `SQPIT-764`
  - [ ] answers the question: _If merged, this PR will: ..._ ³

- The **PR Description**
  - [ ] is free of optional paragraphs and you have filled the relevant parts to the best of your ability
----

# What's new in this PR?

One of many PRs adding support for MLS conference calling

---

Add support for joining a sub conversation as described here: [Use case: Join conference sub-conversation](https://wearezeta.atlassian.net/wiki/spaces/ENGINEERIN/pages/692027483/Use+case+Join+conference+sub-conversation)

In order to join or start an MLS conference call the client must join a sub conversation of the conversation where the call is taking place. The intent is that only clients participating in the call is a member of the sub conversation which is used to generate a shared key for encrypting audio/video streams.

- Add JoinSubconversationUseCase
- Implement subconversation network APIs in V3
- Add incomplete support for starting a MLS conference call

#### Test Coverage (Optional)

- [x] I have added automated test to this contribution

#### How to Test

Not functional yet but the code paths can be trigged by starting a call via the CLI app in a MLS conversation.

----
#### PR Post Submission Checklist for internal contributors (Optional)

 - [ ] Wire's Github Workflow has automatically linked the PR to a JIRA issue
----
#### PR Post Merge Checklist for internal contributors

 - [ ] If any soft of configuration variable was introduced by this PR, it has been added to the relevant documents and the CI jobs have been updated.
----
##### References
1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.

### CODE REVIEW

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/call/usecase/StartCallUseCase.kt [hunk 1]
```diff
@@ -52,7 +54,20 @@ class StartCallUseCase internal constructor(
             conversationType = conversationType,
             isAudioCbr = isAudioCbr
         )
-        Result.Success
+
+        if (conversationType == ConversationType.ConferenceMls) {
+            // TODO update AVS on current epoch after joining the sub conversation
+            joinSubconversationUseCase(conversationId, "conference").fold(
```
Comment (mchenani) [inline logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/call/usecase/StartCallUseCase.kt:60]: maybe we can get rid of this hardcoded text?

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/conversation/JoinSubconversationUseCase.kt [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+    override suspend operator fun invoke(conversationId: ConversationId, subconversationId: String): Either<CoreFailure, Unit> =
+        joinOrEstablishSubconversationAndRetry(conversationId, subconversationId)
+    private suspend fun joinOrEstablishSubconversation(
+        conversationId: ConversationId,
+        subconversationId: String
+    ): Either<CoreFailure, Unit> =
+        wrapApiRequest {
+            conversationApi.fetchSubconversationDetails(conversationId.toApi(), subconversationId)
+        }.flatMap { subconversationDetails ->
+            if (subconversationDetails.epoch > 0UL) {
```
Comment (mchenani) [inline logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/conversation/JoinSubconversationUseCase.kt:43]: let's define something for this 0UL epoch base value :D

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/conversation/JoinSubconversationUseCase.kt [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+                val durationSinceLastEpoch = (subconversationDetails.epochTimestamp?.toInstant()?.timeElapsedUntilNow()?.inWholeHours ?: 0)
```
Comment (mchenani) [inline logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/conversation/JoinSubconversationUseCase.kt:44]: maybe an extension function to do that?

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/conversation/JoinSubconversationUseCase.kt [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+    private suspend fun joinOrEstablishSubconversationAndRetry(
+        conversationId: ConversationId,
+        subconversationId: String
+    ): Either<CoreFailure, Unit> =
+        joinOrEstablishSubconversation(conversationId, subconversationId)
+            .flatMapLeft { failure ->
+                if (failure is NetworkFailure.ServerMiscommunication && failure.kaliumException is KaliumException.InvalidRequestError) {
+                    if (failure.kaliumException.isMlsStaleMessage()) {
+                        kaliumLogger.w("Epoch out of date for conversation $conversationId, re-fetching and re-trying")
+                        // Try again
```
Comment (mchenani) [inline logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/conversation/JoinSubconversationUseCase.kt:85]: we can get rid for this comment, the log in above is explanatory enough.

**Review by mchenani — CHANGES_REQUESTED** [review]
added some comments

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/conversation/JoinSubconversationUseCase.kt (hunk 3 above)
Comment (typfel) [reply logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/conversation/JoinSubconversationUseCase.kt:44]: On which Class? I already added one on `Instant`. But I agree that line is very long :) Maybe just unwrapping the Instant?

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/call/usecase/StartCallUseCase.kt (hunk 1 above)
Comment (typfel) [reply logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/call/usecase/StartCallUseCase.kt:60]: Yep definitely

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/conversation/JoinSubconversationUseCase.kt (hunk 2 above)
Comment (typfel) [reply logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/conversation/JoinSubconversationUseCase.kt:43]: Is it really necessary / helpful to define a constant for 0?

Comment (mchenani) [reply logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/conversation/JoinSubconversationUseCase.kt:43]: I get your point, but  might be helpful for other reviewers to understand what is 0!

**Review by mchenani — APPROVED** [review]
Solid and clean as always 🏄‍♂️

**Pull request merged**